packages/next/src/shared/lib/router/action-queue.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  isThenable,
+  type AppRouterState,
+  type ReducerActions,
+  type ReducerState,
+  ACTION_REFRESH,
+  ACTION_SERVER_ACTION,
+  ACTION_NAVIGATE,
+} from '../../../client/components/router-reducer/router-reducer-types'
+import type { ReduxDevToolsInstance } from '../../../client/components/use-reducer-with-devtools'
+import { reducer } from '../../../client/components/router-reducer/router-reducer'
+import React, { startTransition } from 'react'
+import { createEmptyCacheNode } from '../../../client/components/app-router'
+
+export type DispatchStatePromise = React.Dispatch<ReducerState>
+
+export type AppRouterActionQueue = {
+  state: AppRouterState | null
+  devToolsInstance?: ReduxDevToolsInstance
+  dispatch: (payload: ReducerActions, setState: DispatchStatePromise) => void
+  action: (state: AppRouterState, action: ReducerActions) => ReducerState
+  pending: ActionQueueNode | null
+  needsRefresh?: boolean
+  last: ActionQueueNode | null
+}
+
+export type ActionQueueNode = {
+  payload: ReducerActions
+  next: ActionQueueNode | null
+  resolve: (value: PromiseLike<AppRouterState> | AppRouterState) => void
+  reject: (err: Error) => void
+  discarded?: boolean
+}
+
+export const ActionQueueContext =
+  React.createContext<AppRouterActionQueue | null>(null)
+
+function runRemainingActions(
+  actionQueue: AppRouterActionQueue,
+  setState: DispatchStatePromise
+) {
+  if (actionQueue.pending !== null) {
+    actionQueue.pending = actionQueue.pending.next
+    if (actionQueue.pending !== null) {
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      runAction({
+        actionQueue,
+        action: actionQueue.pending,
+        setState,
+      })
+    }
+  }
+}
+
+async function runAction({
+  actionQueue,
+  action,
+  setState,
+}: {
+  actionQueue: AppRouterActionQueue
+  action: ActionQueueNode
+  setState: DispatchStatePromise
+}) {
+  const prevState = actionQueue.state
+  if (!prevState) {
+    // This shouldn't happen as the state is initialized in the dispatcher if it's not set
+    throw new Error('Invariant: Router state not initialized')
+  }
+
+  actionQueue.pending = action
+  actionQueue.last = action
+
+  const payload = action.payload
+  const actionResult = actionQueue.action(prevState, payload)
+
+  function handleResult(nextState: AppRouterState) {
+    // if we discarded this action, the state should also be discarded
+    if (action.discarded) {
+      if (actionQueue.needsRefresh) {
```

Comment:
I think it's sufficient to just have one refresh at the very end. So maybe this code should only run once the queue is empty?